coding/javascript.md

```diff
@@ -37,6 +37,28 @@ We use [Yarn](https://yarnpkg.com/en) for dependency management.
 
 We use the [Jest](https://jestjs.io) unittest framework.
 
+#### General Testing Advice
+* Take care when using `beforeEach`. It should be clear from reading a test in isolation what state (e.g. fixtures, mocks) it is in. Use your best judgement, but be aware that using `beforeEach` can create a number of issues:
+  * The developer must look in more than one place to understand the test.
+  * As the test suite grows, developers may add fixtures to `beforeEach` that are only used by a subset of tests, and then the onnus is on the developer to understand whether that fixture is relevant for the test they are reading.
+  * A developer may want a fixture in a *similar* state to one already defined in `beforeEach` and chooses to mutate it in their test, reducing clarity. 
+* Don't test external code/libraries.
```

Comment:
It seems like we have very little 'General Testing Advice'.

While "Don't test external code/libraries" is exactly what I would expect to see in this section, first point about `beforeEach` seems quite specific and detailed.

I think that this section would ideally start with and have more points like "Don't test external code/libraries" and maybe later have more implementation detailed advice like the one about `beforeEach`.

One point that we could add here maybe is some advice about what to test? That we should focus on testing public APIs, functions exported by modules, but we don't need to have a unit test of every internal function?
Something like: "Write tests for public APIs of your modules/classes/components. Avoid unit testing internal functions to make tests independent from implementation details."?

Reply:
> One point that we could add here maybe is some advice about what to test? That we should focus on testing public APIs, functions exported by modules, but we don't need to have a unit test of every internal function?
Something like: "Write tests for public APIs of your modules/classes/components. Avoid unit testing internal functions to make tests independent from implementation details."?

Sounds good!

Reply:
That does sound good, but how about proposing it in a follow up branch so that we can land something? This doesn't need to be perfect, and we can iterate on it.

Reply:
> Write tests for public APIs of your modules/classes/components. Avoid unit testing internal functions to make tests independent from implementation details

I don't agree with that advice - unit testing isn't just about correctness, but a tool to help you find the correct design. You should absolutely write tests for internal logic.

Reply:
Just the question if components are a small enough unit, or actually sth that calls for integration tests.

I have dabbled with internal tests before, since no public APIs are available for internal functions.
But I am happy to approve this, and iterate on it.